query-engine/connectors/sql-query-connector/src/ordering.rs

```diff
@@ -1,120 +1,199 @@
 use quaint::ast::*;
 
 static ORDER_JOIN_PREFIX: &str = "orderby_";
 static ORDER_AGGREGATOR_ALIAS: &str = "orderby_aggregator";
+static ORDER_RELEVANCE_VALUE_ALIAS: &str = "orderby_relevance";
 
 #[derive(Debug, Clone)]
-pub struct OrderingJoins {
-    // Joins necessary to perform the order by
+pub struct OrderByDefinition {
+    /// Final column identifier to be used for the scalar field to order by
+    pub(crate) order_column: Expression<'static>,
+    /// Defines ordering for an `ORDER BY` statement.    
+    pub(crate) order_definition: OrderDefinition<'static>,
+    /// Joins necessary to perform the order by
     pub(crate) joins: Vec<AliasedJoin>,
-    // Final column identifier to be used for the scalar field to order by
-    pub(crate) order_column: Column<'static>,
+    /// Top-level column that needs to be selected
+    pub(crate) column_to_select: Option<Expression<'static>>,
 }
 
 /// Builds all expressions for an `ORDER BY` clause based on the query arguments.
 #[tracing::instrument(skip(query_arguments, base_model))]
 pub fn build(
     query_arguments: &QueryArguments,
     base_model: &ModelRef, // The model the ordering will start from
-) -> (Vec<OrderDefinition<'static>>, Vec<OrderingJoins>) {
+) -> Vec<OrderByDefinition> {
     let needs_reversed_order = query_arguments.needs_reversed_order();
 
-    let mut order_definitions: Vec<OrderDefinition<'static>> = vec![];
-    let mut ordering_joins = vec![];
-
     // The index is used to differentiate potentially separate relations to the same model.
-    for (index, order_by) in query_arguments.order_by.iter().enumerate() {
-        let (joins, order_column) = compute_joins(order_by, index, base_model);
-        let order = Some(order_by.sort_order.into_order(needs_reversed_order));
-
-        if joins.is_empty() && order_by.sort_aggregation.is_some() {
-            match order_by.sort_aggregation.unwrap() {
-                SortAggregation::Count => {
-                    order_definitions.push((count(order_column.clone()).into(), order));
-                }
-                SortAggregation::Avg => {
-                    order_definitions.push((avg(order_column.clone()).into(), order));
-                }
-                SortAggregation::Sum => {
-                    order_definitions.push((sum(order_column.clone()).into(), order));
-                }
-                SortAggregation::Min => {
-                    order_definitions.push((min(order_column.clone()).into(), order));
-                }
-                SortAggregation::Max => {
-                    order_definitions.push((max(order_column.clone()).into(), order));
-                }
+    query_arguments
+        .order_by
+        .iter()
+        .enumerate()
+        .map(|(index, order_by)| match order_by {
+            OrderBy::Scalar(order_by) => build_order_scalar(order_by, base_model, needs_reversed_order, index),
+            OrderBy::Aggregation(order_by) if order_by.is_scalar_aggregation() => {
```

Comment:
The `is_scalar_aggregation` strikes me as odd. Why can't this be transformed during parse time?